Single configurable Dockerfile with multi-stage build + external bazel cache

vidlb opened the pull request:
In response to issue #36 , this PR should considerably ease the docker build process, mostly thanks to build arguments and bazel remote cache. You can now build different docker images with custom base image, versions / branches using build arguments, the Dockerfile should be compatible with any branch. You can control env / flags and apt dependencies in external dedicated files.

Thanks to bazel remote cache you can build many docker images without having to rebuild TF (I believe you're going to LOVE it, bazel is such a nightmare !). Now you could just stop and restart a build without loosing progress !
You can choose either CPU or GPU with the only mandatory argument (BASE_IMG), cuda support is enabled if it is found in /usr/local/cuda-*.

I also added some little extras, like passwordless sudo, GUI-related packages option, CPU_RATIO arg to control number of jobs, args to keep TF and/or OTB sources, RTX 3000 support (CUDA compute capability 8.6) and more useful options - see the multibuild.sh script and README.md for available arguments and examples.
All files are now installed in /opt/otbtf (pip wheel as well, in /opt/otbtf/lib/python3.8).

Several changes were made in order to build with latest TF 2.4 and CUDA 11 drivers (ubuntu20.04 with python3.8, no more --noincompatible_do_not_split_linking_cmdline option, bazel and protobuf versions bumps), and since TF 2.3 (I believe) model path is tensorflow::tstring instead of std::string , so I made the fix in order to build, and ran some tests with my old CNN workflow, but I'm not sure this is enough, you may need to apply that fix elsewhere in the repo ? check [this commit](https://github.com/LaTeleScop/otbtf/commit/daa8057b68ce4b850c169ea265f31ef5576f1b80).
You can pull from our [gitlab repo](https://gitlab.com/latelescop/docker/otbtf/container_registry/913118) if you want to check it out !

Mission accomplished, it significantly reduced images size (in this example without GUI and OTB sources only for -dev) :
r1.7 is with the other, more complicated 'bruteforce' method.
Div by 2 for GPU, by 3 for CPU !
![Screenshot_2020-12-30 Container Registry · La TeleScop Docker OTBTF](https://user-images.githubusercontent.com/34066246/103374917-078d6e00-4ad9-11eb-9ddc-6a2bed0f3d33.png)

P.S. 
Sorry for the mess with all those small commits ... I just push too easily and there were so many typos ^^

remicres commented:
Hi @vidlb I just saw you MR. It is just amazing ! Can't wait to test that tomorrow !

vidlb commented:
@remicres haaa finally =D
Hope it builds ! Please tell me how it goes... I didn't try with v2.2 or 2.3 but it should also work, with the right bazel version !

What I found really confusing is the protobuf version ! With v2.4, pip installed python protoc v3.14, but it seems I needed to build v3.9.2 in order to run TF... WTF ?

Also I hope the numpy version problem is not really one. This is the only thing that could require to modify the dockerfile, in order to set a maximum version when installing it with pip.

But won't change the fact that GDAL is installed with numpy as dependencies, with an old version, so I guess it is a quite a mess if you try to import gdal, tf and otbApplication at the same time....

remicres commented:
GPU image built,uncompressed 15Gb instead of 22Gb, well done!
I am giving a try with the CPU image.
I guess when you push the image in the registry, it is compressed again?

vidlb commented:
Yes it is compressed just before upload.
You still need a lot of space in /var/lib/docker if you want to keep all layers, docker will store layers for each different build tag...
See ` docker system df` to check how big it is. When finished there are several ways to clean it, just try `docker builder prune -a` I went with the more brutal `docker system prune` ...

Now the magic with remote cache is that you could run system prune without the need to rebuild tf ! 
BUT I noticed the slightest change in   env / bazel config / tensorflow git tag    will skip the cache (may be just used for some downloads but none of the built files)...

remicres commented:
Well this is a great contribution. Least I can say!
I would be dumb not to merge that in otbtf asap. You are still working on the MR, I am waiting your go. Also, I think we can remove the old docker files, completely obsolete now... Many thanks Vincent

vidlb commented:
Okay, I think I will push some last comments, and remove the dockerfiles dir !

I wasn't sure about that but I can confirm now, they (the TF dev team) did not build and test cuda11.1 yet... So 11.0 should be more stable for now. 
But there was no ubuntu 20 docker image with both cuda11.1 and cudnn8.
So the other options are `11.0-cudnn8-devel-ubuntu18.04` (tf >= 2.4) and `10.2-cudnn7-devel-ubuntu18.04` for older branches.

 I know it because I tried to build with TensorRT and faced [this issue](https://github.com/tensorflow/tensorflow/issues/46173#issuecomment-759182588). I built with compute capabilities 8.6 but we can't be sure that TF is working OK with RTX 3000...

remicres commented:
Hm I don't know. I'd say: Let stick to some supported versions for now, i.e. cuda 11.0 / ubuntu 18.04 ?

vidlb commented:
Yes it is safer !
I'm building with 11.0 right now. Seems ok but I had to remove compute capability 8.6.

By the way, I found a lot of duplicated files : I don't know if it's due to bazel > 1.0, but it seems every include files are installed by pip in /opt/otbtf/lib/python3.*/site-packages/tensorflow/include . So we don't need to manually copy all those file in /opt/otbtf/include after bazel build.

I'll try to identify which lines may be removed... I believe all those are useless : 
`
 && cp -r tensorflow/cc /opt/otbtf/include/tensorflow \
 && cp -r tensorflow/core /opt/otbtf/include/tensorflow \
 && cp -r third_party /opt/otbtf/include \
 && cp -r bazel-tensorflow/external/eigen_archive/unsupported /opt/otbtf/include \
 && cp -r bazel-tensorflow/external/eigen_archive/Eigen /opt/otbtf/include \
 && cp -r tensorflow/lite/tools/make/downloads/absl/absl /opt/otbtf/include \
`
Those directories also contains .cc files, and some build artifacts, we don't need them in order to build the module, do we ?

remicres commented:
Indeed, in principle we only need headers from `include`, not `.cc` files!

vidlb commented:
Okay, I removed all those `cp` lines... and every headers are installed by pip, except one file !
`tensorflow/cc/saved_model/tag_constants.h` is missing in the wheel.
I just copied it and OTBTF built successfully.

vidlb commented:
Now you can merge it ! I downgraded default versions, and was able to remove some unnecessary steps : 
- duplicated headers copy (now just one file)
- the `./tensorflow/lite/tools/make/download_dependencies.sh` and    `./tensorflow/lite/tools/make/build_lib.sh ` lines which are probably related to tensorflow-lite ?
- the protobuf build : I guess it is build by bazel, we don't need to take care of it.

Those images are now smaller (~12gb uncompressed for gpu-dev, 5gb compressed).

I tried with XLA but it causes the same error than with CUDA11.1.
`Check failed: stream->parent()->GetBlasGemmAlgorithms(&algorithms)`.

So it may be useless to build with XLA enabled. Or may be we're missing something in conf.

vidlb commented:
Also I made a little commit to your python files (added a shebang) in order to link it in PATH in the docker images, so we can call it directly from any directory (without the explicit python call and file path).

vidlb commented:
Now you can merge it ! I downgraded default versions, and was able to remove some unnecessary steps :

    duplicated headers copy (now just one file)
    the ./tensorflow/lite/tools/make/download_dependencies.sh and ./tensorflow/lite/tools/make/build_lib.sh lines which are probably related to tensorflow-lite ?
    the protobuf build : I guess it is build by bazel, we don't need to take care of it.

Those images are now smaller (~12gb uncompressed for gpu-dev, 5gb compressed).

I tried with XLA but it causes the same error than with CUDA11.1.
Check failed: stream->parent()->GetBlasGemmAlgorithms(&algorithms).

So it may be useless to build with XLA enabled. Or may be we're missing something in conf.

Also I made a little commit to your executable python files (added a shebang) in order to link it in PATH in the docker images, so we can call it directly from any directory (without the explicit python call and file path).

remicres commented:
I believe that there is an issue with WSL (windows) when XLA is not enabled (see #24).
Have checked how images work on Windows?

vidlb commented:
Nope but I can try, with docker desktop

vidlb commented:
Okay, just don't merge for now, there may be some other things to investigate, also for the cpu build. I'll let you know if it works ok.

remicres commented on tools/docker/build-env-tf.sh:
```diff
@@ -0,0 +1,42 @@
+export TF_NEED_CUDA=0
+export CUDA_TOOLKIT_PATH=$(find /usr/local -maxdepth 1 -type d -name 'cuda-*')
+if  [ ! -z $CUDA_TOOLKIT_PATH ] ; then
+    export LD_LIBRARY_PATH="$LD_LIBRARY_PATH:$CUDA_TOOLKIT_PATH/lib64:$CUDA_TOOLKIT_PATH/lib64/stubs"
```
This makes me think that you have kept the fix provided in #24 (see @Pratyush1991 comment [here](https://github.com/remicres/otbtf/issues/24#issuecomment-658538475)) so it's great new!

remicres commented on tools/docker/build-env-tf.sh:
```diff
@@ -0,0 +1,42 @@
+#export CC_OPT_FLAGS="-march=native -Wno-sign-compare"
+export PYTHON_BIN_PATH=$(which python)
+export PYTHON_LIB_PATH="$($PYTHON_BIN_PATH -c 'import site; print(site.getsitepackages()[0])')"
+export TF_ENABLE_XLA=1
```
Great. Indeed XLA is required (see [here](https://github.com/remicres/otbtf/issues/24#issuecomment-658538475))

Pratyush1991 commented:
Hello @vidlb , 

Thank you for this awesome work. I tried it on my WSL2 Windows. I built the docker with cuda 11 and XLA enabled (by default). The command that I used was as per the readme section.
`docker build --network='host' -t otbtf:gpu-dev --build-arg BASE_IMG=nvidia/cuda:11.0-cudnn8-devel-ubuntu18.04 --build-arg KEEP_SRC_OTB=true .`   
I would like to confirm that it works perfectly fine on WSL2. I'm not sure why you got the error with Cuda 11 and XLA enabled during model serve and I could generate output as well.

vidlb commented:
Hi @Pratyush1991 
The error I had was only ony Linux. I'm not even if sure XLA is supposed to work if running on UNIX. Or may be the TF team hasn't finished to test and patch tf2.4 with XLA... may be we should try with 2.3.2 just to be sure.
But if it works OK on WSL2 this is great !

@remicres  I will soon push my last commits (with a few more improvements regarding the KEEP_SRC_... arguments + now keeping OTBTF sources in /src/otbtf ).
Also I switched OTB and TF to git tags by default in order to avoid bazel rebuild if someone committed (on the branch r2.4 in my case).

vidlb commented:
@remicres I just realized that bazel produce 3 differents libs : libtensorflow_cc.so +  libtensorflow_cc.so.2 +  libtensorflow_cc.so.2.4.0 . Same thing for libtensorflow_framework.
Those are not symbolic links, I checked with sha256sum and they are exactly the same files, this is awkward.
For the GPU build one lib is +400 MB so it seems we could save 800MB more here !

Pratyush1991 commented:
@vidlb  The issue that I have mentioned in #28 still open as I checked on the new build as well. I see someone has posted a fix but I'm not sure I can understand how to implement it. I know it's not related to this may be but do you think anything that can be done for now to fix it. 
Related issues [1](https://github.com/google/or-tools/issues/1830), [2](https://stackoverflow.com/questions/37051635/several-shared-object-using-same-proto-leading-the-the-error-file-already-exist) , [3](https://github.com/protocolbuffers/protobuf/issues/4126), [4](https://github.com/protocolbuffers/protobuf/issues/1489)

vidlb commented:
I'm not sure, but the numpy versions problem could explain this.
Since we installed gdal with apt, we get an outdated np pkg from ubuntu's repo.
Could be something else, but if this is the source of the error, it would require to build gdal (and deps) from source with the SuperBuild, so we can make sure GDAL / OTB and TF are built using the same numpy version.

This would make the build significantly longer..
May be I'll give it a try in a few days.

vidlb commented:
@remicres I found something interesting regarding #28 : 
On my system I have /opt/otb and /opt/tensorflow . 
I can import otb (which I built against libtensorflow_cc )  without my TF env variables (PATH and LD_LIBRARY_PATH).

I tried the test command : 
`Python 3.8.5 (default, Sep  4 2020, 07:30:14)
[GCC 7.3.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import otbApplication
>>> print(otbApplication.Registry_GetAvailableApplications())
2021-01-18 20:09:44.499243: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.11.0
`
So even if I don't import tensorflow, it seems the lib is loaded anyway by OTB. So my guess is that the tensorflow lib is linked to otbApplication, so when we import otbApplication and tensorflow, it is loaded twice in memory, thus the protobuf error.

remicres commented:
@Pratyush1991 Thanks for your feedback! It is great news that this works fine on WSL2.

Indeed, #28 it is still open. I concede that me too, I didn't understand the fix they used in deepmind/reverb... 

I did reproduce the issue using @vidlb docker images, but I think we should address this issue in a different MR. This issue isn't related to docker: we can reproduce it in any debian-like environment for instance.

vidlb commented:
Finally ! You can merge these `42` commits =)

To resume last changes :
-  TF tag 2.4.1 + CUDA 11.0 with ubuntu20.04 (found it !)
- Added arg NUMPY_SPEC="~=1.19"
- Remove bazel and BAZEL_VERSION, added bazelisk
- Use libtensorflow_framework.so.2 installed with the wheel + added `cp -P` to avoid 3 copies of the lib
- Added cuda compute capability 7.0 for Tesla V100 / TITAN V (cc 8.6 / RTX 3000 should drop with TF2.5)
- Re-added `-march=native`, images may not run with old CPUs (need AVX2)  - default build without MKL
- Link python executables in /opt/otbtf/bin
- Copy repo in /src/otbtf
- More python import tests

Also I tried to keep bazel cache when KEEP_SRC_TF but impossible to rebuild using the bazel cache stored inside the container. It goes again from scratch. So right now this argument will only save tensorflow repo (in my list called dev-all). 
I tested on linux and everything should be good.

Images are now little lighter. You could just pull, tag and push those images to dockerhub if you don't want to build it all !

![Screenshot_20210122_181412](https://user-images.githubusercontent.com/34066246/105525015-c1df5200-5ce0-11eb-85db-02a58f609f3f.jpg)
